src/seq_join.rs

```diff
@@ -0,0 +1,356 @@
+use futures::{stream::TryCollect, Future, Stream, TryStreamExt};
+use pin_project::pin_project;
+use std::{
+    collections::VecDeque,
+    future::IntoFuture,
+    iter::Fuse,
+    num::NonZeroUsize,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+/// This helper function might be necessary to convince the compiler that
+/// the return value from [`seq_try_join_all`] implements `Send`.
+/// Use this if you get higher-ranked lifetime errors that mention `std::marker::Send`.
+pub fn assert_send<'a, O>(
+    fut: impl Future<Output = O> + Send + 'a,
+) -> impl Future<Output = O> + Send + 'a {
+    fut
+}
+
+/// Sequentially join futures from an iterator.
+///
+/// This function polls futures from a stream in strict sequence.
+/// If any future blocks, up to `active - 1` futures after it will be polled so
+/// that they make progress.
+///
+/// Unlike [`StreamExt::buffered`], Futures from the stream must resolve in the
```

Comment:
this requirement is interesting, I read the code and I believe it is coming from the implementation where at the time when future is polled and resolved out of order, we don't have a container to store the value received from it. Given that we box anyway, we could probably use an enum to describe elements within the deque: either future or value received from it. 

I don't think we need to do it in this PR, just wondering if it is something we should consider in the future. I don't think we would see impact for in-memory implementations, but curious if this could help us avoid crashes when we start doing real IO

Reply:
This might be the way we resolve some of the weird ordering problems I'm dealing with, yeah.  The random bit generator (may it burn in a very special place) has this property and maybe switching types would help there.  I'll put that into a follow-up (just as I want to follow up with the change to using a `Stream` at some point).